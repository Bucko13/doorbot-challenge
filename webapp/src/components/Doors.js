import React from 'react';
import PropTypes from 'prop-types';
import { Icon, Dimmer, Grid, Button, Header, Segment } from 'semantic-ui-react'

const Doors = ({ list, onOpen }) => {
  const available_doors = list.map((door) => (
    <Grid.Column width={3} key={door.id} data-test={`door-${door.id}`}>
      <Dimmer.Dimmable
        as={Segment}
        compact
        stacked
        blurring
        dimmed={door.opened}
        textAlign='center'>
        <Dimmer active={door.opened} inverted>
          <Header as='h3' icon data-test={`door-${door.id}-opened`}>
            <Icon name='unlock' />
            Opened!
          </Header>
        </Dimmer>
        <Header as='h3'>{door.name}</Header>
        <Button
          data-test={`door-${door.id}-button`}
          onClick={() => onOpen(door)}
          color='teal'
          size='tiny'
        >
          Open
          </Button>
      </Dimmer.Dimmable>
    </Grid.Column>
  ));

  return (
    <Segment size='tiny' padded='very'>
      <Header as='h1'>Doors control</Header>
      { !list.length ?
        <p data-test='message'>There are no doors available for you right now.</p> :
        <p data-test='message'>You have access to these {list.length} doors:</p>
      }
      <Grid stackable stretched>
        <Grid.Row data-test='doors'>
          { available_doors }
        </Grid.Row>
      </Grid>
    </Segment>
  )
}

Doors.propTypes = {
  list: PropTypes.array.isRequired,
  onOpen: PropTypes.func.isRequired,
};

export default Doors;
