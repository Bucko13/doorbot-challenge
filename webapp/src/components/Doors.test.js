import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Doors from './Doors';

describe('Doors Component', () => {
  it('starts with no doors to display', () => {
    const wrapper = generateDoor();
    const doors = wrapper.find(sel('doors'));
    const message = wrapper.find(sel('message'));

    expect(doors.children().length).toBe(0);
    expect(message.text()).toBe('There are no doors available for you right now.');
  });

  it('displays N doors', () => {
    const wrapper = generateDoor({ list: [{ id: 1 }, { id: 2 }, { id: 3 }] });
    const doors = wrapper.find(sel('doors')); const message = wrapper.find(sel('message'));
    const door = wrapper.find(sel('door-1'));

    expect(doors.children().length).toBe(3);
    expect(toJson(door)).toMatchSnapshot();
    expect(message.text()).toBe('You have access to these 3 doors:');
  });

  describe('when loaded with doors', () => {
    let wrapper;
    const onOpen = jest.fn()

    beforeEach(() => {
      wrapper = generateDoor({
        list: [{ id: 1 }, { id: 2, opened: true }, { id: 3 }],
        onOpen,
      });
    });

    it('opens a door', () => {
      const door = wrapper.find(sel('door-1-button'));
      door.simulate('click');
      expect(onOpen.mock.calls.length).toBe(1);
    });

    it('shows the door is opened', () => {
      const doorOpened = wrapper.find(sel('door-2-opened'));
      expect(doorOpened.render().text()).toBe('Opened!');
      expect(toJson(doorOpened)).toMatchSnapshot();
    });

  });

})

function generateDoor(props = {}) {
  const propsToUse = {
    list: [],
    onOpen: () => {},
    ...props,
  }

  return shallow(<Doors {...propsToUse} />);
}

function sel(id) {
  return `[data-test="${id}"]`;
}
