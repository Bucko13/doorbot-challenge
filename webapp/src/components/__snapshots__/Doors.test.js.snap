// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doors Component displays N doors 1`] = `
<GridColumn
  data-test="door-1"
  key="1"
  width={3}
>
  <DimmerDimmable
    as={[Function]}
    blurring={true}
    compact={true}
    stacked={true}
    textAlign="center"
  >
    <Dimmer
      inverted={true}
    >
      <Header
        as="h3"
        data-test="door-1-opened"
        icon={true}
      >
        <Icon
          as="i"
          name="unlock"
        />
        Opened!
      </Header>
    </Dimmer>
    <Header
      as="h3"
    />
    <Button
      as="button"
      color="teal"
      data-test="door-1-button"
      onClick={[Function]}
      size="tiny"
    >
      Open
    </Button>
  </DimmerDimmable>
</GridColumn>
`;

exports[`Doors Component when loaded with doors shows the door is opened 1`] = `
<Header
  as="h3"
  data-test="door-2-opened"
  icon={true}
>
  <Icon
    as="i"
    name="unlock"
  />
  Opened!
</Header>
`;
